rts/motoko-rts/src/idl.rs

```diff
@@ -488,3 +516,344 @@ unsafe extern "C" fn skip_fields(tb: *mut Buf, buf: *mut Buf, typtbl: *mut *mut
         *n -= 1;
     }
 }
+
+unsafe fn is_opt_reserved(typtbl: *mut *mut u8, end: *mut u8, t: i32) -> bool {
+    if is_primitive_type(t) {
+        return t == IDL_PRIM_reserved;
+    }
+
+    // unfold t
+    let mut t = t;
+
+    let mut tb = Buf {
+        ptr: *typtbl.add(t as usize),
+        end: end,
+    };
+
+    t = sleb128_decode(&mut tb);
+
+    return t == IDL_CON_opt;
+}
+
+// TODO: consider storing fixed args typtbl1...end2 in `rel` to use less stack.
+unsafe fn sub(
+    rel: &BitRel,
+    p: bool,
+    typtbl1: *mut *mut u8,
+    typtbl2: *mut *mut u8,
+    end1: *mut u8,
+    end2: *mut u8,
+    t1: i32,
+    t2: i32,
+) -> bool {
+    if t1 >= 0 && t2 >= 0 {
+        let t1 = t1 as u32;
+        let t2 = t2 as u32;
+        if rel.visited(p, t1, t2) {
+            // visited? (bit 0)
+            // return assumed or determined result
+            return rel.related(p, t1, t2);
+        };
+        // cache and continue
+        rel.visit(p, t1, t2); // mark visited
+        rel.assume(p, t1, t2); // assume t1 <:/:> t2 true
+    };
+
+    /* primitives reflexive */
+    if is_primitive_type(t1) && is_primitive_type(t2) && t1 == t2 {
+        return true;
+    }
+
+    // unfold t1, if necessary
+    let mut tb1 = Buf {
+        ptr: if t1 < 0 {
+            end1
+        } else {
+            *typtbl1.add(t1 as usize)
+        },
+        end: end1,
+    };
+
+    let u1 = if t1 >= 0 {
+        sleb128_decode(&mut tb1)
+    } else {
+        t1
+    };
+
+    // unfold t2, if necessary
+    let mut tb2 = Buf {
+        ptr: if t2 < 0 {
+            end2
+        } else {
+            *typtbl2.add(t2 as usize)
+        },
+        end: end2,
+    };
+
+    let u2 = if t2 >= 0 {
+        sleb128_decode(&mut tb2)
+    } else {
+        t2
+    };
+
+    // NB we use a trivial labelled loop so we can factor out the common failure continuation.
+    // exit either via 'return true' or 'break 'return_false' to memoize the negative result
+    'return_false: loop {
+        match (u1, u2) {
+            (_, IDL_CON_alias) | (IDL_CON_alias, _) => idl_trap_with("sub: unexpected alias"),
+            (_, IDL_PRIM_reserved) => return true,
+            (IDL_PRIM_empty, _) => return true,
+            (IDL_PRIM_nat, IDL_PRIM_int) => return true,
+            (_, IDL_CON_opt) => return true, // apparently, this is admissable
```

Comment:
these all return `true`, why not have an or-pattern?

Reply:
Will do